File: Recursion/PowerSet of Array/PowerSet.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;

class PowerSet {

    static void subsetHelper(int i, ArrayList<Integer> subset, ArrayList<Integer> nums, ArrayList<ArrayList<Integer>> ans) {
    
        if(nums.size() == i){
            ans.add(new ArrayList<>(subset));
            return;
        }
    
        subset.add(nums.get(i));
        subsetHelper(i+1, subset, nums, ans);
        subset.remove(subset.size() - 1);
        subsetHelper(i+1, subset, nums, ans);
        return;
    }
   
    static ArrayList<ArrayList<Integer>> powerSet(ArrayList<Integer> nums){
        ArrayList<ArrayList<Integer>> ans  = new ArrayList<>();
        ArrayList<Integer> subset = new ArrayList<Integer>();
        subsetHelper(0, subset, nums, ans);
        return ans;
    }

   static void arrayElements(ArrayList<Integer> arr){
    int len = arr.size();
    System.out.print("{ ");
    for(int i = 0; i<len; i++) {
        System.out.print(arr.get(i));
        if(i!=len-1){
            System.out.print(", ");
            }
        }
        System.out.print(" }\n");
    }
    public static void main(String args[]){
        ArrayList<ArrayList<Integer>> powerset;
        System.out.print("Enter number of elements in power set: ");
        Scanner input = new Scanner(System.in);
        ArrayList<Integer> arr = new ArrayList<Integer>();
        int number = input.nextInt();
        for(int i=0; i<number; i++) {
            System.out.println("Enter number "+ (i+1) +" : ");
            int value = input.nextInt();
            arr.add(value);
        }
        System.out.print("Array elements: ");
        arrayElements(arr);
        System.out.print("Power set: \n");
        powerset = powerSet(arr);
        for(int i=0; i<powerset.size(); i++) {
            arrayElements(powerset.get(i));
        }
        input.close();
    }
}

File: Recursion/006.PowerSet of Array/PowerSet.swift
//
//  PowerSet
//
//  Created by Feialoh Francis on 23/11/22.
//

import Foundation

class PowerSet {
    
    var arr: [Int]
    
    init(arr: [Int]) {
        self.arr = arr
    }
    
    func helper (_ i: Int, _ subset: [Int], _ ans: [[Int]]) -> [[Int]] {
        var ans = ans
        var subset = subset
        if arr.count == i {
            ans.append(subset)
            return ans
        }
    
        subset.append(arr[i])
        ans = helper(i+1, subset, ans)
        subset.remove(at: subset.count-1)
        ans = helper(i+1, subset, ans)
        return ans
    }
    
    func powerSet() ->  [[Int]] {
        return helper(0, [Int](), [[Int]]())
    }
    
    func reverse(_ l: Int, _ r: Int) -> [Int] {
        var arr = self.arr
        if l > r {
            return arr
        }
        arr.swapAt(l, r)
        self.arr = arr
        return reverse(l+1, r-1)
    }
    
    func reverseArray(_ pos: Int) -> [Int] {
        return reverse(0, self.arr.count-1)
    }
}

var powerSet = [[Int]]()
var arr = [Int]()
print("Enter number of elements in power set: ")
let num = Int(readLine() ?? "0")

if let num = num {
    for i in 0...num-1 {
        print("Enter number \(i+1): ")
        let value = Int(readLine() ?? "0")
        if let element = value {
            arr.append(element)
        }
    }
    
    print("Array elements: \(arr)")
    let sol = PowerSet(arr: arr)
    powerSet = sol.powerSet()
    print("Power set: \(powerSet)")
}
